[Navigation API] dispose-after-bfcache.html and entries-after-bfcache.html fail on WPT

RupinMittal opened the pull request:
#### 855821352f744e79f32c3246ba82010994d965a3
<pre>
[Navigation API] dispose-after-bfcache.html and entries-after-bfcache.html fail on WPT
<a href="https://bugs.webkit.org/show_bug.cgi?id=299918">https://bugs.webkit.org/show_bug.cgi?id=299918</a>
<a href="https://rdar.apple.com/161694156">rdar://161694156</a>

Reviewed by Chris Dumez.

When a restoring a Page from the bfcache, we must also restore the Navigation
object&apos;s list of HistoryItems.

This happens in CachedPage::restore. For each frame, we get all of its HistoryItems
from the UIProcess and send them to Navigation::updateForReactivation. But the
Navigation API&apos;s list can only contain sequential items of the same origin as the
current item.

For example, consider this UI Process b/f list:

1. foo.com
2. bar.com
3. foo.com
4. bar.com
5. bar.com#1  &lt;-- currentItem
6. bar.com#2
7. foo.com
8. bar.com

The Navigation object&apos;s list must be: { 4, 5, 6 } in that order and nothing more.

This means that once we get the b/f list from the UI Process, we must filter it
before giving it Navigation::updateForReactivation.

BackForwardController::reachableItemsForFrame claimed to have this logic, but it&apos;s
logic was wrong. It was not filtering out non-same-origin entries, but rather
getting *all* of the items across a frame&apos;s lifetime.

To fix this, we remove BackForwardController::reachableItemsForFrame and replace it
with Navigation::filterHistoryItemsForNavigationAPI (since this filtering is specific
to the Navigation API). We also refactor the Navigation API restoring code in
CachedPage::restore to be cleaner.

This change will make dispose-after-bfcache.html and entries-after-bfcache.html
pass on WPT. I have manually verified this. They already pass locally. They were
failing on WPT because when the test is run on WPT, an extra item is present in
the b/f list initially, This item wasn&apos;t getting filtered out and was causing the
test to fail. This item is not present when the test is run locally, and so the
test is passing. To simulate this difference, we add a new test:

LayoutTests/http/tests/navigation-api/different-origin-entries-removed-after-bfcache.html

* LayoutTests/http/tests/navigation-api/different-origin-entries-removed-after-bfcache-expected.txt: Added.
* LayoutTests/http/tests/navigation-api/different-origin-entries-removed-after-bfcache.html: Added.
* LayoutTests/http/tests/navigation-api/resources/page-that-goes-back.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::allItems):
(WebCore::BackForwardController::reachableItemsForFrame): Deleted.
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):
(WebCore::CachedPage::restoreNavigationAPIHistoryItems):

For each frame, obtain send the filtered list to the Navigation object.

* Source/WebCore/history/CachedPage.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::filterHistoryItemsForNavigationAPI):

Filter the list in the way described above.

* Source/WebCore/page/Navigation.h:

Canonical link: <a href="https://commits.webkit.org/301283@main">https://commits.webkit.org/301283@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/989170ac613b5fd3337058aaf7e84865a687ff45

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125453 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45120 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35861 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132311 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77341 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0e0ac848-4f08-49eb-8346-42c8f55a1486) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127329 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45807 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53682 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95533 "Failed to checkout and rebase branch from PR 51604") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/66cd52c5-74be-44bc-a193-a3b09231af5b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128406 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36604 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112191 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76059 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/no-web-process-leak (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/299fa140-c021-4b1c-b7ec-d9237a71baee) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35504 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30364 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75787 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106375 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30591 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134991 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52255 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40033 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104009 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52694 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108407 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103757 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49124 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27422 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49419 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19652 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52153 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57937 "") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51504 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54860 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53199 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cdb446a8be8242209a989db3f3da3ade155a6472. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51604

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/64239b5898070e88076d14f9e629cfcd36f01e86. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51604

RupinMittal commented on LayoutTests/platform/wk2/TestExpectations:
```diff
@@ -814,7 +814,8 @@ webkit.org/b/288851 imported/w3c/web-platform-tests/fetch/http-cache/invalidate.
-http/wpt/navigation-api/navigation-activation/activation-after-bfcache.html [ Failure ]
-http/wpt/navigation-api/navigation-history-entry/entries-after-bfcache-in-iframe.html [ Failure ]
-http/wpt/navigation-api/navigation-history-entry/entries-after-bfcache.html [ Failure ]
+webkit.org/b/99917 http/wpt/navigation-api/navigation-activation/activation-after-bfcache-cross-origin.html [ Failure ]
```
Added a test.

cdumez commented on LayoutTests/http/tests/navigation-api/different-origin-entries-removed-after-bfcache-expected.txt:
```diff
@@ -0,0 +1,7 @@
+PASS successfullyParsed is true
+
+TEST COMPLETE
```
If `TEST COMPLETE` shows in the middle of the test and not at the end then the test is not correctly written.

cdumez commented on LayoutTests/http/tests/navigation-api/different-origin-entries-removed-after-bfcache.html:
```diff
@@ -0,0 +1,26 @@
+<script src="/js-test-resources/js-test.js"></script>
+
+<script>
+testRunner?.dumpAsText();
```
Not needed if you import `js-test.js`

cdumez commented on LayoutTests/http/tests/navigation-api/different-origin-entries-removed-after-bfcache.html:
```diff
@@ -0,0 +1,26 @@
+
+<script>
+testRunner?.dumpAsText();
+testRunner?.waitUntilDone();
```
should use `jsTestIsAsync = true` when using `js-test.js`

cdumez commented on LayoutTests/http/tests/navigation-api/different-origin-entries-removed-after-bfcache.html:
```diff
@@ -0,0 +1,26 @@
+        window.navigation.addEventListener('navigate', function() {
+            debug('Visiting Page 1 for the second time');
+            shouldBeEqualToNumber("window.navigation.entries().length", 1);
+            testRunner?.notifyDone();
```
Should use `finishJSTest()` when using `js-test.js`

cdumez commented on LayoutTests/http/tests/navigation-api/resources/page-that-goes-back.html:
```diff
@@ -0,0 +1,16 @@
+            debug('page-that-goes-back loaded, going back using history.back()');
+            setTimeout(() => {
+                history.back();
+            }, 1000);
```
1 second is a very long time for a test to run, why doesn't `0` suffice?

cdumez commented on Source/WebCore/history/CachedPage.cpp:
```diff
@@ -215,6 +190,26 @@ void CachedPage::restore(Page& page)
+    }
+
+    for (RefPtr child = frame.tree().firstChild(); child; child = child->tree().nextSibling()) {
+        if (RefPtr localChild = dynamicDowncast<LocalFrame>(child.get()))
```
`*child` to avoid an unnecessary null check.

cdumez commented on Source/WebCore/page/Navigation.h:
```diff
@@ -185,6 +185,8 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
 
     void updateNavigationEntry(Ref<HistoryItem>&&, ShouldCopyStateObjectFromCurrentEntry);
 
+    static Vector<Ref<HistoryItem>> filterHistoryItemsForNavigationAPI(Vector<Ref<HistoryItem>>&, HistoryItem&);
```
Make up your mind, does it modify the Vector inline (in which case no return value) or does it return the modified vector (in which case the parameter should be `const`.

cdumez replied:
Or, given that your implementation moves things out of the vector, likely we want `Vector<Ref<HistoryItem>>&&`.

cdumez commented on Source/WebCore/history/CachedPage.h:
```diff
@@ -66,6 +67,8 @@ class CachedPage final : public CanMakeCheckedPtr<CachedPage> {
     void markForContentsSizeChanged() { m_needsUpdateContentsSize = true; }
 
 private:
+    void restoreNavigationAPIHistoryItems(LocalFrame&, CheckedRef<BackForwardController>);
```
The parameter should not be a `CheckedRef` here, this doesn't match our Safer CPP guidelines. Please pass in a raw pointer.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -1205,4 +1205,41 @@ void Navigation::abortOngoingNavigationIfNeeded()
+// https://html.spec.whatwg.org/multipage/browsing-the-web.html#getting-session-history-entries-for-the-navigation-api
+Vector<Ref<HistoryItem>> Navigation::filterHistoryItemsForNavigationAPI(Vector<Ref<HistoryItem>>& allItems, HistoryItem& currentItem)
+{
+    std::optional<size_t> startingIndex = std::nullopt;
```
`= std::nullopt` does nothing here.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -1205,4 +1205,41 @@ void Navigation::abortOngoingNavigationIfNeeded()
+Vector<Ref<HistoryItem>> Navigation::filterHistoryItemsForNavigationAPI(Vector<Ref<HistoryItem>>& allItems, HistoryItem& currentItem)
+{
+    std::optional<size_t> startingIndex = std::nullopt;
+    for (size_t i = 0; i < allItems.size(); i++) {
```
`++i`

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -1205,4 +1205,41 @@ void Navigation::abortOngoingNavigationIfNeeded()
+    Vector<Ref<HistoryItem>> filteredItems;
+    Ref startingOrigin = SecurityOrigin::create(currentItem.url());
+
+    for (int i = (int)*startingIndex - 1; i >= 0; i--) {
```
`--i`. Also, please avoid C-style casts.

cdumez commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -1205,4 +1205,41 @@ void Navigation::abortOngoingNavigationIfNeeded()
+    filteredItems.reverse();
+    filteredItems.append(currentItem);
+
+    for (size_t i = *startingIndex + 1; i < allItems.size(); i++) {
```
`++i`

RupinMittal committed:
[Navigation API] dispose-after-bfcache.html and entries-after-bfcache.html fail on WPT
https://bugs.webkit.org/show_bug.cgi?id=299918
rdar://161694156

Reviewed by Chris Dumez.

When a restoring a Page from the bfcache, we must also restore the Navigation
object's list of HistoryItems.

This happens in CachedPage::restore. For each frame, we get all of its HistoryItems
from the UIProcess and send them to Navigation::updateForReactivation. But the
Navigation API's list can only contain sequential items of the same origin as the
current item.

For example, consider this UI Process b/f list:

1. foo.com
2. bar.com
3. foo.com
4. bar.com
5. bar.com#1  <-- currentItem
6. bar.com#2
7. foo.com
8. bar.com

The Navigation object's list must be: { 4, 5, 6 } in that order and nothing more.

This means that once we get the b/f list from the UI Process, we must filter it
before giving it Navigation::updateForReactivation.

BackForwardController::reachableItemsForFrame claimed to have this logic, but it's
logic was wrong. It was not filtering out non-same-origin entries, but rather
getting *all* of the items across a frame's lifetime.

To fix this, we remove BackForwardController::reachableItemsForFrame and replace it
with Navigation::filterHistoryItemsForNavigationAPI (since this filtering is specific
to the Navigation API). We also refactor the Navigation API restoring code in
CachedPage::restore to be cleaner.

This change will make dispose-after-bfcache.html and entries-after-bfcache.html
pass on WPT. I have manually verified this. They already pass locally. They were
failing on WPT because when the test is run on WPT, an extra item is present in
the b/f list initially, This item wasn't getting filtered out and was causing the
test to fail. This item is not present when the test is run locally, and so the
test is passing. To simulate this difference, we add a new test:

LayoutTests/http/tests/navigation-api/different-origin-entries-removed-after-bfcache.html

* LayoutTests/http/tests/navigation-api/different-origin-entries-removed-after-bfcache-expected.txt: Added.
* LayoutTests/http/tests/navigation-api/different-origin-entries-removed-after-bfcache.html: Added.
* LayoutTests/http/tests/navigation-api/resources/page-that-goes-back.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/history/BackForwardController.cpp:
(WebCore::BackForwardController::allItems):
(WebCore::BackForwardController::reachableItemsForFrame): Deleted.
* Source/WebCore/history/BackForwardController.h:
* Source/WebCore/history/CachedPage.cpp:
(WebCore::CachedPage::restore):
(WebCore::CachedPage::restoreNavigationAPIHistoryItems):

For each frame, obtain send the filtered list to the Navigation object.

* Source/WebCore/history/CachedPage.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::filterHistoryItemsForNavigationAPI):

Filter the list in the way described above.

* Source/WebCore/page/Navigation.h:

Canonical link: https://commits.webkit.org/301283@main
